### PR TITLE
IGNITE-17085 Support persistent case for page-memory-based MV storage

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17085

### CODE REVIEW

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryPartitionStorage.java [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+import org.apache.ignite.internal.util.IgniteCursor;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Abstract implementation of {@link PartitionStorage} based on a {@link BplusTree}.
+ */
+// TODO: IGNITE-16644 Support snapshots.
+abstract class AbstractPageMemoryPartitionStorage implements PartitionStorage {
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryPartitionStorage.java:50]: I don't get this class. PartitionStorage is deprecated, what's the point? Can you please explain?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 2]
```diff
@@ -164,8 +170,113 @@ public void destroy() throws StorageException {
 
     /** {@inheritDoc} */
     @Override
-    public PageMemoryMvPartitionStorage createMvPartitionStorage(int partitionId) {
-        throw new UnsupportedOperationException("Not supported yet");
+    public PersistentPageMemoryMvPartitionStorage createMvPartitionStorage(int partitionId) {
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:173]: Do we need an old method for creating partition? It can just throw UnsupportedOperationException.

Why did you rename partId to partitionId in method that has nothing to do with your code btw? Why was it necessary?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 3]
```diff
@@ -164,8 +170,113 @@ public void destroy() throws StorageException {
...
+                    tableView,
+                    dataRegion.pageMemory(),
+                    versionChainFreeList,
+                    rowVersionFreeList,
+                    versionChainTree,
+                    checkpointTimeoutLock
+            );
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException(
+                    String.format("Error getting or creating partition metadata [tableName=%s, partitionId=%s]", tableView.name(),
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:273]: Not the best description. Just remove the word "metadata" and it'll get better

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java [hunk 4]
```diff
@@ -104,60 +103,43 @@ public void start() {
...
     private TableFreeList createTableFreeList(PageMemory pageMemory) throws IgniteInternalCheckedException {
-        long metaPageId = pageMemory.allocatePage(FREE_LIST_GROUP_ID, FREE_LIST_PARTITION_ID, FLAG_AUX);
-
         return new TableFreeList(
                 FREE_LIST_GROUP_ID,
                 FREE_LIST_PARTITION_ID,
                 pageMemory,
                 PageLockListenerNoOp.INSTANCE,
-                metaPageId,
+                pageMemory.allocatePage(FREE_LIST_GROUP_ID, FREE_LIST_PARTITION_ID, FLAG_AUX),
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:111]: Why did you inline this? Old code was clean enough, and new code became more complicated. I don't understand the necessity.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java [hunk 5]
```diff
@@ -104,60 +103,43 @@ public void start() {
...
-
+    private static VersionChainFreeList createVersionChainFreeList(PageMemory pageMemory) throws IgniteInternalCheckedException {
         return new VersionChainFreeList(
                 FREE_LIST_GROUP_ID,
                 FREE_LIST_PARTITION_ID,
                 pageMemory,
-                reuseList,
                 PageLockListenerNoOp.INSTANCE,
-                metaPageId,
+                pageMemory.allocatePage(FREE_LIST_GROUP_ID, FREE_LIST_PARTITION_ID, FLAG_AUX),
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:124]: Same here

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java [hunk 6]
```diff
@@ -104,60 +103,43 @@ public void start() {
...
-        );
-
         return new RowVersionFreeList(
                 FREE_LIST_GROUP_ID,
                 FREE_LIST_PARTITION_ID,
                 pageMemory,
                 reuseList,
                 PageLockListenerNoOp.INSTANCE,
-                metaPageId,
+                pageMemory.allocatePage(FREE_LIST_GROUP_ID, FREE_LIST_PARTITION_ID, FLAG_AUX),
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:141]: And here. Why?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvStorageIoModule.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
 import org.apache.ignite.internal.storage.pagememory.mv.io.VersionChainMetaIo;
 
 /**
- * {@link PageIoModule} related to {@link PageMemoryMvPartitionStorage} implementation.
+ * {@link PageIoModule} related to {@link AbstractPageMemoryMvPartitionStorage} implementation.
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvStorageIoModule.java:31]: Looks weird when it points to an abstract class. I think we can list 2 MvTableStorage implementations here, would be better.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersionFreeList.java [hunk 8]
```diff
@@ -83,7 +80,7 @@ public RowVersionFreeList(
                 LOG,
                 metaPageId,
                 initNew,
-                pageListCacheLimit,
+                null,
```
Comment (ibessonov) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersionFreeList.java:83]: If this thing doesn't exist anymore then why do we have this "null" argument?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryPartitionStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryPartitionStorage.java:50]: We will get rid of this in a separate IGNITE-17466, but for now, let it be right

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:273]: fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:111]: What's more confusing here?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:124]: What's more confusing here?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:141]: What's more confusing here?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvStorageIoModule.java (hunk 7 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvStorageIoModule.java:31]: fix it

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java [hunk 9]
```diff
@@ -21,114 +21,55 @@
...
 import org.junit.jupiter.api.extension.ExtendWith;
 
+/**
+ * Base test for MV partition storages based on PageMemory.
+ */
 @ExtendWith(ConfigurationExtension.class)
 @ExtendWith(WorkDirectoryExtension.class)
-class PageMemoryMvPartitionStorageTest extends AbstractMvPartitionStorageTest<PageMemoryMvPartitionStorage> {
-    private final PageIoRegistry ioRegistry = new PageIoRegistry();
+abstract class AbstractPageMemoryMvPartitionStorageTest<T extends AbstractPageMemoryMvPartitionStorage> extends
```
Comment (sashapolo) [inline modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:46]: Why does `AbstractMvPartitionStorageTest` have a generic parameter? I was able to remove it from this class and from all other inheritors without any implications

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java [hunk 10]
```diff
@@ -21,114 +21,55 @@
...
+        AbstractMvPartitionStorageTest<T> {
+    protected final PageIoRegistry ioRegistry = new PageIoRegistry();
 
-    private final BinaryRow binaryRow3 = binaryRow(key, new TestValue(22, "bar3"));
+    protected final BinaryRow binaryRow3 = binaryRow(key, new TestValue(22, "bar3"));
```
Comment (sashapolo) [inline modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:50]: can still be private

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:124]: Fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:141]: Fix it

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java [hunk 11]
```diff
@@ -259,9 +210,9 @@ void scanByTimestampWorksCorrectlyAfterCommitAndAbortFollowedByUncommittedWrite(
 
     @Test
     void readByTimestampWorksCorrectlyIfNoUncommittedValueExists() {
-        LinkRowId rowId = storage.insert(binaryRow, txId);
+        LinkRowId rowId = (LinkRowId) insert(binaryRow, txId);
```
Comment (sashapolo) [inline modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:213]: what is this cast for?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersionFreeList.java (hunk 8 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersionFreeList.java:83]: Fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java (hunk 4 above)
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryDataRegion.java:111]: Because it was more clear that this allocated page is a meta page

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java [hunk 12]
```diff
@@ -80,32 +80,89 @@ void testPageCount() {
         assertEquals(2, meta.pageCount());
     }
 
+    @Test
+    void testVersionChainTreeRootPageId() {
+        PartitionMeta meta = new PartitionMeta(null, 0, 0, 0, 0, 0, 0);
+
+        assertEquals(0, meta.versionChainTreeRootPageId());
+
+        assertDoesNotThrow(() -> meta.versionChainTreeRootPageId(null, 100));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java:89]: I'll leave.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 13]
```diff
@@ -164,8 +170,113 @@ public void destroy() throws StorageException {
...
-    public PageMemoryMvPartitionStorage createMvPartitionStorage(int partitionId) {
-        throw new UnsupportedOperationException("Not supported yet");
+    public PersistentPageMemoryMvPartitionStorage createMvPartitionStorage(int partitionId) {
+        TableView tableView = tableCfg.value();
+
+        FilePageStore filePageStore = ensurePartitionFilePageStore(tableView, partitionId);
+
+        CheckpointManager checkpointManager = dataRegion.checkpointManager();
+
+        CheckpointTimeoutLock checkpointTimeoutLock = checkpointManager.checkpointTimeoutLock();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:180]: Maybe.

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java [hunk 14]
```diff
@@ -80,32 +80,89 @@ void testPageCount() {
         assertEquals(2, meta.pageCount());
     }
 
+    @Test
+    void testVersionChainTreeRootPageId() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java:84]: Yes, but also to check metadata snapshots for those setters

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 15]
```diff
@@ -226,21 +336,92 @@ TableFreeList createTableFreeList(
...
+    /**
+     * Returns new {@link VersionChainFreeList} instance for partition.
+     *
+     * @param tableView Table configuration.
+     * @param partId Partition ID.
+     * @param rootPageId Root page ID.
+     * @param initNew {@code True} if new metadata should be initialized.
+     * @throws StorageException If failed.
+     */
+    VersionChainFreeList createVersionChainFreeList(
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:348]: It seems to me that the factory is more suitable when we do not know the exact type of the object, but here we do.
Made them private

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java (hunk 10 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:50]: fix it

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java (hunk 11 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:213]: Because the method returns `RowId`

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java (hunk 9 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:46]: Fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:179]: Can you create a ticket for fixing this?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java]: I suggested a factory here, because it will allow to extract some code from this class. We have a tendency to have very bloated classes.

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java (hunk 11 above)
Comment (sashapolo) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:212]: And why do you need a `LinkRowId`?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:179]: I don't know because we're trying to take the lock by timeout.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java]: I propose to leave it as it is, the factory will slightly reduce the code, but it can complicate it.

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java (hunk 11 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorageTest.java:212]: Fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java]: I disagree, but ok

**Review by sashapolo — APPROVED** [review]

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java [hunk 16]
```diff
@@ -37,7 +37,7 @@
 public class PartitionMetaTest {
     @Test
     void testTreeRootPageId() {
-        PartitionMeta meta = new PartitionMeta(null, 0, 0, 0);
+        PartitionMeta meta = new PartitionMeta(null, 0, 0, 0, 0, 0, 0);
```
Comment (ibessonov) [inline modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java:40]: This constructor starts looking ridiculous, maybe we should use some default constructor and only set fields that we need?

Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaTest.java]: fix it
